### PR TITLE
Escape characters in CSV

### PR DESCRIPTION
**Pull request by emartynov**
It is fix for #383.

Changes:
1. Refactored CSV export
2. Escaped '\n', '\r', '"', ',' when generating CSV
3. Extracted one more string constant
4. Added some missing tests

No changes here:
1. Using async task
2. Resume export after user allows write permission

Please review and comment.

### CODE REVIEW

**Comment by emartynov** [comment]
@AndyScherzinger can you review?

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java [hunk 1]
```diff
@@ -22,110 +22,105 @@
...
-            return null;
+        } else {
+            for (GlucoseReading reading : readings) {
+                writeLine(osw,
+                        this.dateTool.convertRawDate(reading.getCreated()),
+                        this.dateTool.convertRawTime(reading.getCreated()),
+                        formatter.format(GlucosioConverter.glucoseToMmolL(reading.getReading())),
+                        Constants.Units.MMOL_L,
+                        valueOrEmptyString(reading.getReading_type()),
+                        valueOrEmptyString((reading.getNotes())
```
[the remark this answers is not in the document]
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java]: Fixed

File: app/src/main/java/org/glucosio/android/presenter/ExportPresenter.java [hunk 2]
```diff
@@ -34,166 +35,144 @@
...
                     super.onPostExecute(filename);
-                    if (filename != null) {
-                        Uri uri = FileProvider.getUriForFile(mActivity.getApplicationContext(),
-                                mActivity.getApplicationContext().getPackageName() + ".provider.fileprovider", new File(filename));
-                        mExportView.onExportFinish(uri);
-                    } else if (isEmpty[0]) {
-                        mExportView.onNoItemsToExport();
+                    if (EMPTY.equals(filename)) {
+                        view.onNoItemsToExport();
+                    } else if (filename != null) {
```
[the remark this answers is not in the document]
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/presenter/ExportPresenter.java]: Ok, applied

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java]: Fixed

File: app/src/main/java/org/glucosio/android/presenter/ExportPresenter.java [hunk 3]
```diff
@@ -34,166 +35,144 @@
...
-
-    public int getFromYear() {
-        return fromYear;
-    }
+            new ReadingToCSV(context, preferredUnit).createCSVFile(readings, osw);
 
-    public void setFromYear(int fromYear) {
-        this.fromYear = fromYear;
+            return file.getPath();
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/presenter/ExportPresenter.java]: Fixed

**Comment by AndyScherzinger** [comment]
@emartynov sure, I'll give the branch a test drive :)

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java [hunk 4]
```diff
@@ -22,110 +22,104 @@
...
+    private void writeLine(@NonNull OutputStreamWriter osw, @NonNull String... values) throws IOException {
         for (int i = 0; i < values.length; i++) {
-            osw.append(values[i]);
+            osw.append(escapeCSV(values[i]));
             osw.append(i == values.length - 1 ? '\n' : ',');
         }
     }
+
+    private String escapeCSV(String value) {
+        if (containsSymbolsToEscape(value)) {
```
Comment (AndyScherzinger) [inline app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: ```SYMBOLS_TO_ESCAPE``` actually contains more symbols that get checked (and lead to a ```return true```) than the following line will actually escape.

**Comment by AndyScherzinger** [comment]
Looks good to me code wise, just a minor comment. Will test the export later today after work.

**Comment by AndyScherzinger** [comment]
Like mentioned in the in line comment only " and ' seem to be escaped, line breaks still exist and won't be escaped. This leads to new lines which will be treated as new rows, e.g. the following export:
```
Date,Time,Concentration,Preferred Glucose unit,Measured,Notes
1/22/18,21:37,25,mg/dL,After dinner,"Yet
Another
'
""
Test"
1/22/18,21:36,23.5,mg/dL,After dinner,"This, is
A
New line
Test"
```
When imported into MS Excel (komma delimiter, "-String key):
2018-01-22 21_53_03-mappe1 - excel

So I would guess an in-app import of it would still fail (don't know how to do an import :/)

**Comment by emartynov** [comment]
We don't have import, only export :(

I was quickly checking https://tools.ietf.org/html/rfc4180. It looks like MS Excel is wrong with importing CSV.

I will dive more :)

**Comment by AndyScherzinger** [comment]
I also send it to my gmail adress and opened it as preview within gmail and there it works fine... so yeah, maybe we just have to ignore Excel...

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 4 above)
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: Can you elaborate more? I don't get the comment, sorry.

Did I make some logical error?

**Comment by emartynov** [comment]
Conclusion on this PR?

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 4 above)
Comment (AndyScherzinger) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: I'll try :) containsSymbolsToEscape checks for a certain set of characters while not all of them get escaped then.

**Comment by AndyScherzinger** [comment]
@emartynov fine by me and as discussed excel has its cvs issues but that is excel not glucose :)

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 4 above)
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: Can you be more specific? What characters are not escaped?

Comment (AndyScherzinger) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: I'll try :)

```
 if (containsSymbolsToEscape(value)) {
    return "\"" + value.replace("\"", "\"\"") + "\"";
} 
```
escaping is done for ```"``` while the condition ```containsSymbolsToEscape(value)``` needs to be met of course. Thus I expected the condition to check for the same pattern. But it actually checks for:
```
private boolean containsSymbolsToEscape(String value) {
        for (char c : SYMBOLS_TO_ESCAPE) {
            if (value.indexOf(c) > -1) {
                return true;
            }
        }

        return false;
    }
```
while ```SYMBOLS_TO_ESCAPE``` contains:
```
private static final char[] SYMBOLS_TO_ESCAPE = new char[]{',', '"', '\n', '\r'};
```` 
So I would have expected the actual ascaping to also then escape the other "symbols to escape".

Comment (AndyScherzinger) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: other than that also when ```\n``` has been found the check return true and you are escaping for ```"```

**Comment by AndyScherzinger** [comment]
Hi @emartynov I tried to explain it again, sorry for not getting it clearly explained :( Hope this time I got it right :)

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 4 above)
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: If `,` or `\n` or `\r` is detected then escaping is wrapping the whole string in the single quotes `"`. However, if the string contains single quote it should be escaped as double quote `""`. I hope it helps.

Comment (AndyScherzinger) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: Ah 👍 
I didn't think of that, now makes send. Thanks for the explanation and your patience 😃

**Comment by AndyScherzinger** [comment]
All fine by me. Any feedback @paolorotolo?

File: app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java (hunk 4 above)
Comment (emartynov) [reply app/src/main/java/org/glucosio/android/tools/ReadingToCSV.java:109]: No issues, I also didn't know it by heart and should, of course, mention it in PR description.

**Comment by paolorotolo** [comment]
I'll try it ASAP

**Comment by paolorotolo** [comment]
Works fine in LibreOffice, good work

screenshot from 2018-01-29 15-35-52

**Review by paolorotolo — APPROVED** [review]

**Pull request merged**